File: elixir/bob/bob.exs
defmodule Bob do
  defp is_shouting(input) do
    input == String.upcase(input)
    && String.downcase(input) != String.upcase(input)
  end

  defp is_question(input) do
    String.ends_with?(input, "?")
  end

  defp is_silence(input) do
    input |> String.trim |> String.length == 0
  end

  def hey(input) do
    cond do
      is_silence(input)
        -> "Fine. Be that way!"
      is_shouting(input) && is_question(input)
        -> "Calm down, I know what I'm doing!"
      is_shouting(input)
        -> "Whoa, chill out!"
      is_question(input)
        -> "Sure."
      true
        -> "Whatever."
    end
  end
end

File: php/bob/bob.php
<?php

class Bob {

  function respondTo($gambit)
  {
    $gambit = $this->mb_trim($gambit);

    if ($this->isSilence($gambit)) {
      return 'Fine. Be that way!';
    }

    if ($this->isShouting($gambit) && $this->isQuestion($gambit)) {
      return "Calm down, I know what I'm doing!";
    }

    if ($this->isShouting($gambit)) {
      return 'Whoa, chill out!';
    };

    if ($this->isQuestion($gambit)) {
      return 'Sure.';
    }

    return 'Whatever.';
  }

  private function isShouting($utterance)
  {
    return $utterance === mb_strtoupper($utterance)
        && $utterance !== mb_strtolower($utterance);
  }

  private function isQuestion($utterance)
  {
    return mb_substr($utterance, -1) === '?';
  }

  private function isSilence($utterance)
  {
    return mb_strlen($utterance) === 0;
  }

  private function mb_trim($str) {
    return preg_replace('/^[\p{Z}\s]+|[\p{Z}\s]+$/u','',$str);
  }
}
